controllers/btpoperator_controller.go

```diff
@@ -424,75 +220,178 @@ func (r *BtpOperatorReconciler) HandleProcessingState(ctx context.Context, cr *v
+	return nil
 }
 
-func (r *BtpOperatorReconciler) createUnstructuredObjectsFromManifestsDir(ctx context.Context, manifestsDir string) ([]*unstructured.Unstructured, error) {
+func (r *BtpOperatorReconciler) deleteOutdatedResources(ctx context.Context) error {
 	logger := log.FromContext(ctx)
 
+	logger.Info("getting outdated module resources to delete")
+	usToDelete, err := r.createUnstructuredObjectsFromManifestsDir(r.getResourcesToDeletePath())
+	if err != nil {
+		logger.Error(err, "while getting deletable objects from manifests")
+		return fmt.Errorf("Failed to create deletable objects from manifests: %w", err)
+	}
+	logger.Info(fmt.Sprintf("got %d outdated module resources to delete", len(usToDelete)))
+
+	err = r.deleteResources(ctx, usToDelete)
+	if err != nil {
+		logger.Error(err, "while deleting outdated resources")
+		return fmt.Errorf("Failed to delete outdated resources: %w", err)
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) createUnstructuredObjectsFromManifestsDir(manifestsDir string) ([]*unstructured.Unstructured, error) {
 	objs, err := r.manifestHandler.CollectObjectsFromDir(manifestsDir)
 	if err != nil {
-		logger.Error(err, "while collecting objects from manifests")
 		return nil, err
 	}
 	us, err := r.manifestHandler.ObjectsToUnstructured(objs)
 	if err != nil {
-		logger.Error(err, "while converting objects to Unstructured")
 		return nil, err
 	}
 
 	return us, nil
 }
 
-func (r *BtpOperatorReconciler) prepareModuleResources(ctx context.Context, us []*unstructured.Unstructured) *ErrorWithReason {
+func (r *BtpOperatorReconciler) getResourcesToDeletePath() string {
+	return fmt.Sprintf("%s%cdelete", ResourcesPath, os.PathSeparator)
+}
+
+func (r *BtpOperatorReconciler) deleteResources(ctx context.Context, us []*unstructured.Unstructured) error {
+	logger := log.FromContext(ctx)
+
+	var errs []string
+	for _, u := range us {
+		if err := r.Delete(ctx, u); err != nil && !k8serrors.IsNotFound(err) {
+			errs = append(errs, fmt.Sprintf("failed to delete %s %s: %s", u.GetName(), u.GetKind(), err))
+		} else {
+			logger.Info("deleted resource", "name", u.GetName(), "kind", u.GetKind())
+		}
+	}
+
+	if errs != nil {
+		return errors.New(strings.Join(errs, ", "))
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) reconcileResources(ctx context.Context, s *corev1.Secret) error {
+	logger := log.FromContext(ctx)
+
+	logger.Info("getting module resources to apply")
+	usToApply, err := r.createUnstructuredObjectsFromManifestsDir(r.getResourcesToApplyPath())
+	if err != nil {
+		logger.Error(err, "while creating applicable objects from manifests")
+		return fmt.Errorf("Failed to create applicable objects from manifests: %w", err)
+	}
+	logger.Info(fmt.Sprintf("got %d module resources to apply", len(usToApply)))
+
+	logger.Info("preparing module resources to apply")
+	if err = r.prepareModuleResources(ctx, usToApply, s); err != nil {
+		logger.Error(err, "while preparing applicable object to apply")
```

Comment:
Nit again: applicable = those that can be applied. In this context this is pleonasm.

Reply:
changed to `...preparing objects to apply`